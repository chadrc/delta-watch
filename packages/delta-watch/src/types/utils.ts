import {ObjectWatcher} from "../core/ObjectWatcher";

export interface DeltaWatchInternals {
  watcher: ObjectWatcher
  type: string
}

export function makeAccessorHandler(mutatorMethods: string[]) {
  return {
    get: function (obj: Date, prop: PropertyKey) {
      if (prop in obj) {
        let field = (obj as any)[prop];
        if (typeof field === 'function') {
          if (mutatorMethods.indexOf(prop as string) !== -1) {
            throw Error("Cannot access a mutator method on an Accessor object.");
          }
          field = field.bind(obj);
        }
        return field;
      }
    },
    set: function () {
      throw Error("Cannot set a value on the Accessor object.");
    }
  };
}

export type CanSetPropFunc = (prop: PropertyKey) => boolean;

export function makeMutationHandler<T extends object>(internals: DeltaWatchInternals,
                                                      mutatorMethods: string[],
                                                      canSetProp: boolean | CanSetPropFunc = false ): ProxyHandler<T> {
  return {
    get: function (_: T, prop: PropertyKey) {
      if (prop === "__DeltaWatchInternals") {
        return internals;
      }

      let fieldMutator = ObjectWatcher.getMutator(internals.watcher, prop);
      let fieldWatcher = ObjectWatcher.getFieldWatcher(internals.watcher, prop);
      if (prop in internals.watcher._data) {
        let field = (internals.watcher._data as any)[prop];
        if (typeof field === 'function') {
          field = field.bind(internals.watcher._data);
          if (mutatorMethods.indexOf(prop as string) !== -1) {
            return function (...args: any[]) {
              let result = field(...args);
              internals.watcher._notifySubscribers(true, true, true);
              return result;
            }
          }
        } else {
          // let fieldMutator = ObjectWatcher.getMutator(self._watcher, field);
          // return fieldMutator == null ? self._watcher._properties[field]._data : fieldMutator;

          if (fieldMutator !== null && typeof fieldMutator !== 'undefined') {
            return fieldMutator;
          } else if (fieldWatcher !== null && typeof fieldWatcher !== 'undefined') {
            internals.watcher._makeMutator(prop);
            return ObjectWatcher.getMutator(internals.watcher, prop);
          }
        }
        return field;
      } else if (fieldMutator !== null && typeof fieldMutator !== 'undefined') {
        return fieldMutator;
      } else if (fieldWatcher !== null && typeof fieldWatcher !== 'undefined') {
        internals.watcher._makeMutator(prop);
        return ObjectWatcher.getMutator(internals.watcher, prop);
      }
    },
    set: function (_: T, prop: PropertyKey, value: any): boolean {
      if (prop === "__DeltaWatchInternals") {
        throw Error("Cannot set value of __DeltaWatchInternals");
      }

      let willSetProp;
      if (canSetProp instanceof Function) {
        willSetProp = canSetProp(prop);
      } else {
        willSetProp = canSetProp;
      }
      // If any prop should be accepted as settable
      // or settableProps contains the given prop
      if (willSetProp === true) {
        // Need to mutate before array check,
        // because watcher's make mutator uses current value to determine
        // to make Object or Array mutator

        // if data is undefined then we need to make it an object
        if (typeof internals.watcher._data === 'undefined') {
          internals.watcher._parent._data[internals.watcher._parentKey] = {};
        }
        internals.watcher._data[prop] = value;
        let fieldMutator = ObjectWatcher.getMutator(internals.watcher, prop);
        let type = internals.watcher._typeRegistry.getTypeForValue(value);

        if (Array.isArray(value) &&
          (fieldMutator === null || typeof fieldMutator === 'undefined'
            || (fieldMutator as any).__DeltaWatchInternals.type !== type)) {
          // if mutator hasn't been initialized
          // or the type of the mutator doesn't match the new value
          // make a new mutator for it
          internals.watcher._makeMutator(prop);
        }

        let watcherProperties = internals.watcher._properties[prop];
        if (watcherProperties) {
          watcherProperties._notifySubscribers(true, true);
        }
        return true;
      }

      throw Error(`${prop} not settable on Mutator of type ${internals.type}`)
    }
  };
}