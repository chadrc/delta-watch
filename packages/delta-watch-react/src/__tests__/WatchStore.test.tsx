import 'jest';
import * as React from 'react';
import DeltaWatchReact from '../index';
import * as renderer from 'react-test-renderer';
import './setup'
import {flushTimers} from "./utils";

describe(`WatchStore`, () => {
  it(`Mutating a watchable store causes a wrapped component to re-render`, () => {
    let {WatchStore, Store} = DeltaWatchReact.MakeStore({
      message: "Hello"
    });

    let renderCount = 0;

    class Text extends React.Component<any, any> {
      render() {
        renderCount++;
        return (
          <p>{this.props.message}</p>
        );
      }
    }

    let Wrapped = WatchStore((watcher: any) => ({
      message: watcher.message
    }))(Text);

    let component = renderer.create(<Wrapped/>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    Store.Mutator.message = "Delta Watch Render";

    flushTimers();

    tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    // 1 for initial render, 1 for re-render
    expect(renderCount).toEqual(2);
  });

  it(`Multiple mutations to store only cause one update`, () => {
    let {WatchStore, Store} = DeltaWatchReact.MakeStore({
      headline: "Multiple Updates",
      message: "No Updates"
    });

    let renderCount = 0;

    class Text extends React.Component<any, any> {
      render() {
        renderCount++;
        return (
          <div>
            <h1>{this.props.headline}</h1>
            <p>{this.props.message}</p>
          </div>
        );
      }
    }

    let Wrapped = WatchStore((watcher: any) => ({
      headline: watcher.headline,
      message: watcher.message
    }))(Text);

    let component = renderer.create(<Wrapped/>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    Store.Mutator.headline = "Updated";
    Store.Mutator.message = "One Update";

    flushTimers();

    tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    // 1 for initial render, 1 for re-render
    expect(renderCount).toEqual(2);
  });

  it(`Providing a mapStore function transforms props`, () => {
    let {WatchStore, Store} = DeltaWatchReact.MakeStore({
      items: [
        'item one',
        'item two'
      ],
      selectedItem: 0
    });

    class Text extends React.Component<any, any> {
      render() {
        return (
          <div>
            <p>{this.props.selectedItem}</p>
          </div>
        );
      }
    }

    let Wrapped = WatchStore(
      (watcher: any) => ({
        selectedItem: watcher.selectedItem
      }),
      (accessor: any, props: any) => ({
        selectedItem: accessor.items[props.selectedItem]
      })
    )(Text);

    let component = renderer.create(<Wrapped/>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    Store.Mutator.selectedItem = 1;

    flushTimers();

    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it(`Getting new props maps to different watch`, () => {
    let {WatchStore, Store} = DeltaWatchReact.MakeStore({
      items: [
        'item one',
        'item two'
      ],
      selectedItem: 0
    });

    const Child = (props) => <p>{props.item}</p>;

    const ChildWatcher = WatchStore((watcher: any, props: any) => ({
      item: watcher.items[props.itemIndex]
    }))(Child);

    const Parent = (props) => (
      <div>
        <ChildWatcher itemIndex={props.selectedItem}/>
      </div>
    );

    const ParentWatcher = WatchStore((watcher: any) => ({
      selectedItem: watcher.selectedItem
    }))(Parent);

    let component = renderer.create(<ParentWatcher/>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    Store.Mutator.selectedItem = 1;

    flushTimers();

    tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    // items[0] should also have no watchers
    expect(Store.Watcher.items[0]._watcherOptions.length).toEqual(0);
  });
});