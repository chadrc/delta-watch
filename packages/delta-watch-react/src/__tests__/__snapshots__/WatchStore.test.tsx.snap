// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WatchStore Getting new props maps to different watch 1`] = `
<div>
  <p>
    item one
  </p>
</div>
`;

exports[`WatchStore Getting new props maps to different watch 2`] = `
<div>
  <p>
    item two
  </p>
</div>
`;

exports[`WatchStore Multiple mutations to store only cause one update 1`] = `
<div>
  <h1>
    Multiple Updates
  </h1>
  <p>
    No Updates
  </p>
</div>
`;

exports[`WatchStore Multiple mutations to store only cause one update 2`] = `
<div>
  <h1>
    Updated
  </h1>
  <p>
    One Update
  </p>
</div>
`;

exports[`WatchStore Mutating a watchable store causes a wrapped component to re-render 1`] = `
<p>
  Hello
</p>
`;

exports[`WatchStore Mutating a watchable store causes a wrapped component to re-render 2`] = `
<p>
  Delta Watch Render
</p>
`;

exports[`WatchStore Providing a mapStore function transforms props 1`] = `
<div>
  <p>
    item one
  </p>
</div>
`;

exports[`WatchStore Providing a mapStore function transforms props 2`] = `
<div>
  <p>
    item two
  </p>
</div>
`;
